### ISSUE TITLE
restrict extension permissions

### ISSUE THREAD

**Issue by limonte** [FlowCrypt/flowcrypt-browser#1897]
I started with the installation of Firefox extension and was expecting the extension to ask for permission to mail.google.com only, but FlowCrypt is asking to access all websites. This can potentially stop users concerned about their online security.

![image](https://user-images.githubusercontent.com/6059356/62848341-e493cb80-bce3-11e9-9e0b-3d07e8ea5d51.png)

For the comparison, here are the permissions asked by https://addons.mozilla.org/en-US/firefox/addon/checker-plus-gmail/
![image](https://user-images.githubusercontent.com/6059356/62848305-ba420e00-bce3-11e9-8b44-0974b498526e.png)

Is it possible to narrow down the asked permissions to gmail-related domains only?

**Comment by limonte** [comment]
The same goes for Chrome extension: 

FlowCrypt: 

![image](https://user-images.githubusercontent.com/6059356/62848703-67695600-bce5-11e9-8851-7bdc275f5086.png)

Checker Plus:

![image](https://user-images.githubusercontent.com/6059356/62848780-b0210f00-bce5-11e9-9674-5b6184cbbba1.png)

**Comment by tomholub** [comment]
Awesome. Yes.

This is because we are planning to bring support for other webmails: Outlook, etc.

If we limited manifest to mail.google.com only, and later added another one, all of our user's extensions will be disabled! And there will be wartime on support.

If we added a limited list gmail/outlook/etc, that obviously also limits us in the future.

Luckily Google comes to the rescue with extension manifest v3. You can find the draft spec (or maybe today finalized?). It allows us to request permissions for "potentially any domain", but not actually receive permission for any domain. That means to scary popup for the user. Then, based on which webmail user uses, we can request permission later for that particular domain, say mail.google.com

So basically we're talking about incremental permissions, which is exactly what we need. Available once we transition to manifest v3. Maybe you could take a look for me about when support is planned to land in Chrome, or potentially firefox if at all? Some links would be helpful for future reference.

**Comment by tomholub** [comment]
(marking actionable, but for research only)

**Comment by tomholub** [comment]
Will be superseded by #1509 and #2158 . Manifest V3 will bring us this flexibility.

**Comment by tomholub** [comment]
until manifest v3 is available, we could restrict to the following list:

```
"https://*.google.com/*",
"https://*.flowcrypt.com/*",
"https://outlook.live.com/*'
"https://outlook.office365.com/*"
"https://outlook.office.com/*"
```
the last one is for `https://outlook.office.com/noop/cryptup` oauth redirect which may possibly change

Over time we could further restrict it as things evolve, or transition to V3 where user has more granual control over how/where the extension will run.

**Comment by tomholub** [comment]
This one has to be tested very extensively, to ensure that the browser will not auto-disable the extension as it may due to permission change.

**Comment by limonte** [comment]
After reading docs and answers on StackOverflow I'm still not sure what will happen when changing permissions for the new release.

I'll go ahead and create a simple extension and reproduce the scenario with changing persmissions in it to see how both browsers will react.

**Comment by tomholub** [comment]
It depends if the browser thinks the permissions were "expanded" or not. In this case, it shouldn't (because we are replacing a wildcard with specific domains). But it needs to be tested.

If it thinks permissions were expanded, it should auto-disable plugin during update.

Not sure if this also applies to extensions loaded from disk directly, you'd need to test that.

**Comment by tomholub** [comment]
I played around with the permissions for a bit

It seems that the highest permission we require is the `tabs` permission. When present, this will show:

![image](https://user-images.githubusercontent.com/6306961/73015848-b5e25b00-3e14-11ea-8415-9f606c07aef6.png)

So we'll have to figure out how to stop using the `tabs` permission, but I'll leave that for another issue.

For this issue, I'll take the risk. I'll update 10% of users first and wait for complaints to roll in, if any. If they don't, we'll update everybody. If they do, we'll re-evaluate and maybe roll back in next update, maybe not.

**Comment by tomholub** [comment]
I figured, we have to lower the permission one way or another, so if that means we have to go through some pain to get there, then we don't have much choice about it anyway (if we do want the updated permissions).

**Comment by tomholub** [comment]
Given that this is important enough to do it regardless if it disables existing users (who'd have to manually re-enable it), knowing the details in advance won't affect the decision, so it's not worth the time to spend examining it further.

**Comment by tomholub** [comment]
Initially, I thought it would be easier to test. But knowing that it's really difficult to test, it's probably a waste of time to try evaluating it further.

**Comment by drzraf** [comment]
I do think that keeping `*.google.com` instead of `mail.google.com` is a frighteningly broad and obviously unnecessary permission.
 `*.google.com` or  `mail.google.com` does not provide any present or future advantage wrt `outlook`.
I suggest to urgently restrict this regexp to the sole `mail.google.com` (as mailvelope does).

**Comment by tomholub** [comment]
It uses more than mail.google.com - also account.google.com for xoauth2. There may be more, I have to check. We have also received requests to add support for google groups and so on. I'll try to enumerate the list of services we could potentially support, and then restrict the permission to those.

I'll make another issue. It won't be addressed urgently, but it will be addressed.

**Comment by tomholub** [comment]
Thanks for the input!
